harness/determined/__init__.py

```diff
@@ -23,6 +24,14 @@
 from determined import errors
 from determined import util
 
+try:
+    # Attempt to get the version from the package metadata. This should
+    # exist if using a built distribution (i.e. for most users), which
+    # includes pip editable installations.
+    __version__ = importlib.metadata.version(__name__)
+except importlib.metadata.PackageNotFoundError:
+    __version__ = "2!0.0.0+unknown"
```

Comment:
I'm baffled about the meaning of 2!0.0.0+unknown - could you enligthen me?

Reply:
It should never be the case that we are unable to return a version string set in the package metadata, because if someone is using the `determined` Python package, it should be some sort of distribution, which will always be built with package metadata. Stranger things have happened, though, so this is the default I return if, heaven forbid, there's no version metadata to read. I slapped a `2!` epoch specifier on the front so this default version would satisfy existing version constraints in `requirements.txt` (the higher epoch orders it after any other versions we could possibly publish, which have an implicit `0` epoch by default).

Reply:
I.e. when you run `import determined; determined.__version__`, Python will return the version string included as part of the distribution metadata. But `__version__` should under no circumstances ever break or throw an exception, so I have this set as a degenerate default.